### PR TITLE
BUG: special: erfinv(x<<1) loses precision

### LINKED ISSUE #12758: BUG: special: erfinv(y) is inaccurate for small abs(y).
`erfinv(y)` is implemented as `ndtri(0.5 * (y+1)) * NPY_SQRT1_2` (where `NPY_SQRT1_2` is `1/sqrt(2)`).  This limits the precision that can be obtained for small `abs(y)`, because for any `y` such that `abs(y)` is less than ~1e-16, `y + 1` is `1`.  So, for example, `erfinv(1e-20)` returns 0, but the correct result is `8.86226925452758e-21`:
```
In [45]: erfinv(1e-20)  # incorrect result                                                          
Out[45]: 0.0

In [46]: erf(8.86226925452758e-21)                                              
Out[46]: 1e-20
```
Octave and Boost return the correct value; we should too.

Octave:
```
octave:3> format long
octave:4> erfinv(1e-20)
ans =    8.86226925452758e-21
```
Boost
```
#include <iostream>
#include <boost/math/special_functions/erf.hpp>

using boost::math::erf_inv;
using namespace std;


int main()
{
    double p = 1e-20;
    double x = erf_inv(p);
    cout << scientific << setprecision(15) << x << endl;
    // 8.862269254527580e-21
}
```

### PR DESCRIPTION
**Pull request by ev-br**
#### Reference issue

closes gh-12758

#### What does this implement/fix?

Avoid the loss of precision for small arguments by using the Taylor series for small x.

#### Additional information

At small arguments, `erf(x) = 2/\sqrt{\pi} (x - x^3 / 3 + ...)`, so we only keep the linear term for `|x| < 1e-7` and then the inverse function `erfinv(x) = x * \sqrt{\pi} / 2`.

### CODE REVIEW

File: scipy/special/tests/test_erfinv.py [hunk 1]
```diff
@@ -57,3 +57,25 @@ def test_literal_values(self):
...
+        expected = np.array([8.86226925452758e-21,
+                             8.862269254527581e-16,
+                             8.86226925452758e-15,
+                             8.862269254527581e-11,
+                             8.86226925452758e-09,
+                             7.97604232907484e-08,
+                             9.74849617998037e-08,
+                             8.8622692545299e-07])
+        assert_allclose(sc.erfinv(x), expected,
+                        rtol=1e-10)
```
[the remark this answers is not in the document]
Comment (ev-br) [reply scipy/special/tests/test_erfinv.py:76]: Theoretically, yes, the threshold of 1e-7 corresponds to the relative error of the order of 1e-14, but then we're at the fiddling with tiny differences territory: https://github.com/scipy/scipy/pull/12760#discussion_r475236299  (esp the last value).
So I'd rather have the tolerance slightly on the loose side.

**Comment by WarrenWeckesser** [comment]
This is a big improvement, but there is still significant loss of precision outside the threshold interval.  For example, at y=2e-7, the relative error is on the order of 1e-10.  With a little more work, we can do a lot better than that.

The tool I usually reach for in a case like this is the [Padé approximant](https://en.wikipedia.org/wiki/Pad%C3%A9_approximant).  It isn't the *optimal* rational approximation (ping @person142!), but it usually works very well, and is easy to generate with the [tools available in `mpmath`](http://mpmath.org/doc/current/calculus/approximation.html). Here's a C function that is based on the (8, 8) Padé approximant of `erfinv` that can be used instead of the linear approximation:

```
/*
 *  Compute erfinv(y) using a Pade approximation.
 *  Tests suggest that on the interval [-0.25, 0.25],
 *  the relative error is roughly 3e-16 or less.
 */
double erfinv_pade_approx(double y)
{
    // Coefficients are in reverse order; the constant term
    // is the last term in the array.
    double P2[] = {-0.001107151790418148,
                   0.05924573283568412,
                   -0.5269332792177829,
                   1.6303793184358268,
                   -2.0404152179484742,
                   0.886226925452758};
    double Q2[] = {-0.006439886366130444,
                   0.1501276021364257,
                   -0.9428722610441058,
                   2.3670502144056558,
                   -2.5641614119568574,
                   1.0};
    double y2 = y*y;
    double num = polevl(y2, P2, 5);
    double den = polevl(y2, Q2, 5);
    return y * (num / den);
}
```
WIth this approximation, the threshold can be increased to 0.25.

That C code was generated using a Python script that I put in a gist: https://gist.github.com/WarrenWeckesser/9d2bd366fd7c1c069c9f9640c548f79d
E.g.
```
In [1]: run erfinv_approx.py                                                                                                           

In [2]: P2, Q2 = create_pade_approx()                                                                                                  

In [3]: generate_c_code(P2, Q2)                                                                                                        
double polevl(double x, const double coef[], int N);

/*
 *  Compute erfinv(y) using a Pade approximation.
 *  Tests suggest that on the interval [-0.25, 0.25],
 *  the relative error is roughly 3e-16 or less.
 */
double erfinv_pade_approx(double y)
{
    // Coefficients are in reverse order; the constant term
    // is the last term in the array.
    double P2[] = {-0.001107151790418148,
                   0.05924573283568412,
                   -0.5269332792177829,
                   1.6303793184358268,
                   -2.0404152179484742,
                   0.886226925452758};
    double Q2[] = {-0.006439886366130444,
                   0.1501276021364257,
                   -0.9428722610441058,
                   2.3670502144056558,
                   -2.5641614119568574,
                   1.0};
    double y2 = y*y;
    double num = polevl(y2, P2, 5);
    double den = polevl(y2, Q2, 5);
    return y * (num / den);
}
```

**Comment by person142** [comment]
If you want an accurate and fast implementation of `erfinv`, you can use this one:

https://github.com/person142/spycial/blob/master/spycial/erfinv.py

(which is ported from Boost, which is using rational approximations via the Remez algorithm).

**Comment by WarrenWeckesser** [comment]
@person142 cool, thanks for the link.

I think I'd rather port directly to C or Cython instead of introducing a dependency on numba.  (Not that numba is bad--I'm just not sure we want the dependency for SciPy.)

But either way, we need to get more Boost into SciPy.  Fixing bugs in cdflib, cephes, etc., is frustrating when you know that Boost has good implementations.

**Comment by person142** [comment]
Oh yeah I'm definitely not recommending we use that implementation directly. Mostly porting from that Python code to C is likely easier than porting from Boost directly because with Boost you have to disentangle the implementations for various different precisions.

**Comment by WarrenWeckesser** [comment]
@person142 you're right, that does look like it would be easier to port to C than the original source.

File: scipy/special/cephes/erfinv.c [hunk 2]
```diff
@@ -17,6 +17,17 @@ double erfinv(double y) {
     const double domain_lb = -1;
     const double domain_ub = 1;
 
+    const double thresh = 1e-7;
```
Comment (sethtroisi) [inline scipy/special/cephes/erfinv.c:20]:
```
import scipy.special
import mpmath
import math

mpmath.dps = 50

M_2_SQRTPI = 2 / math.sqrt(math.pi)

for i in range(1, 20):
    y = 10 ** -i
    correct = float(mpmath.erfinv(y))

    s = scipy.special.erfinv(y)
    s2 = (y / M_2_SQRTPI)

    print (f"{i:2} {(s - correct) / correct:.2e} {(s2 - correct) / correct:.2e}")
```

```
 1 1.09e-15 -2.63e-03
 2 1.17e-15 -2.62e-05
 3 -1.10e-13 -2.62e-07
 4 -1.10e-13 -2.62e-09
 5 6.55e-12 -2.62e-11
 6 -8.23e-11 -2.62e-13
 7 5.84e-10 -2.54e-15
 8 -6.08e-09 0.00e+00
 9 8.27e-08 0.00e+00
...
19 -1.00e+00 0.00e+00
```

I think the single term taylor series generates better results starting from `y < 6e-6`

**Comment by ev-br** [comment]
FWIW I'm unlikely to be able to come back to this PR in the near future. I won't close it since it has nice bits of Pade approximations, which can probably be made into a PR of their own and supersede this one.

**Comment by tylerjereddy** [comment]
@ev-br Based on your comment and the rather complex discussion above, I'll bump the milestone for now--feel free to revert if you'll suddenly have time in the next two weeks.

**Review by tylerjereddy — COMMENTED** [review]
@ev-br @WarrenWeckesser Correct me if I'm wrong here, but as I look at the code and the comments like

> This is a big improvement, but there is still significant loss of precision outside the threshold interval...

I have to ask the simple question--is this not an incremental improvement over current `master` that we could merge and then keep appropriate issues open  to gain even more ground on precision in future PRs?

**Comment by tylerjereddy** [comment]
Ok, I'm going to merge this but keep the cognate issue open since some comments asked for even more precision. If there really is a need for more precision, but the issue is fixed, feel free to close the issue and replace with another for additional precision.

Thanks @ev-br and reviewers.

**Pull request merged**